8364248: Separate memory limit detection

jsikstro opened the pull request:
The function os::has_allocatable_memory() is intended to determine whether there is a system-imposed limit on how much memory can be committed, and if so, what that limit is. On POSIX systems, limiting committable memory is typically enforced by restricting the available virtual address space, such as via RLIMIT_AS. As a result, os::has_allocatable_memory() tells us both how much memory can be committed and how much virtual address space is available. On Windows however, os::has_allocatable_memory() always returns true, along with the size of the available virtual address space. This is misleading because it is not possible to limit how much memory can be committed via virtual address space, and also the virtual address space cannot be limited.

ZGC currently uses os::has_allocatable_memory() to check if the virtual address space is limited. To make it clear that the virtual address space cannot be limited on Windows, I propose that we create a new function called os::has_limited_virtual_address_space() which simply returns false on Windows, since the virtual address space cannot be limited there.

As a follow-up, I think it is reasonable to re-visit the implementation of os::has_allocatable_memory() on Windows, since it doesn't follow any user-set limits, apart from how much virtual memory is available. Perhaps looking at limit(s) set by Job Objects could be more fruitful, and would improve the support for native Windows containers (Hyper-V).

jsikstro commented:
/cc hotspot-gc

tstuefe commented:
> > Generally good. But why not simply return the limit. Unlimited = size_max.
> > Oh, I see ZGC does that already. I would adopt that for has_limited_virtual_address_space. Then you can remove the ZGC implementation completely.
> 
> This would make the ZGC could read more nicely I think. I like it!
> 
> > Side note, to be unbearably nitpicky, it should theoretically be "address space limit - os::vm_min_address()" since the OS does not allow low-address mappings below that.
> 
> Do you think it would be prudent to add this now? To clarify, if there is a limit on POSIX, return `(size_t)rlim.rlim_cur - vm_min_address()`?

Nah, this was a half joke tbh. I don't think this fidelity is needed, and it may confuse people if we get a weirdly unaligned address back. E.g. if on Linux, the min map is configured to be one page, you could get something like "4PB - 4K".

Also, you won't ever be able to use the full extend of that size anyway, since of course it is pre-populated with all kind of mappings, is fragmented, etc.

Side note, it would be valuable to have a sister function to os::vm_min_address() that returns the maximum mappable user address. I did hard code values at various places, but I believe ZGC has some function like this that is more elegant and more truthful. Would be nice to have this in os space.

tstuefe commented:
I have some second thoughts about this. Sorry for the bikeshedding, but naming is important, especially since we never bother with comments on prototypes.

"allocatable_memory_limit" implies much more than it delivers. How much memory you can allocate depends on many factors, only one of which being the user-addressable address space size. 

For example, how much you can *commit* depends on OS, swap space size, OS-specific overcommit behavior etc. How much you can really use depends on how much memory you really have and how the OS is willing to give you. Etc.

I think it would be easier to clearly name this function as "address_space_limit" or similar, since that's what it is doing.

jsikstro commented:
Just want to start of by saying thank you for taking the time with this.

> I have some second thoughts about this. Sorry for the bikeshedding, but naming is important, especially since we never bother with comments on prototypes.
> 
> "allocatable_memory_limit" implies much more than it delivers. How much memory you can allocate depends on many factors, only one of which being the user-addressable address space size.
> 
> For example, how much you can _commit_ depends on OS, swap space size, OS-specific overcommit behavior etc. How much you can really use depends on how much memory you really have and how the OS is willing to give you. Etc.
> 
> I think it would be easier to clearly name this function as "address_space_limit" or similar, since that's what it is doing.

I agree that the naming is strong, but I don't see an issue with that. Like you're saying, on Linux (maybe all POSIX systems?), you can never commit more memory than you have virtual address space, regardless of OS, swap space size, and overcommit behavior. So a strong name for a strong upper-bound seems reasonable to me.

On POSIX, we only really need an os::address_space_limit(), since it answers both how much memory we can commit and the size of the virtual address space. On Windows, we need an os::allocatable_memory_limit(), which would answer how much memory we can commit, which can be adjusted by the user, unlike the size of the virtual address space, which cannot be artificially limited. So I think we need both, so that both can be called from both OS's.

tstuefe commented on src/hotspot/os/posix/os_posix.cpp:
```diff
@@ -790,12 +788,31 @@ bool os::has_allocatable_memory_limit(size_t* limit) {
+  }
+
+  // No limit
+  return SIZE_MAX;
```
This would be a very nice place to hook in the function I proposed you add, the sister function to `os::vm_min_address()`, let's call it `os::vm_max_address()`.

Looking through the ZGC sources, this is the functionality I meant: https://github.com/openjdk/jdk/blob/559795b0eb8061325127fa9fdf8b80617fe47166/src/hotspot/cpu/aarch64/gc/z/zAddress_aarch64.cpp#L45

For platforms that don't have that implementation, you can still return SIZE_MAX as fallback.

jsikstro replied:
Yes, this could be nice to have in the future, and would likely be a larger re-write.

tstuefe commented on src/hotspot/os/posix/os_posix.cpp:
```diff
@@ -713,7 +713,7 @@ bool os::get_host_name(char* buf, size_t buflen) {
 
 #ifndef _LP64
-// Helper, on 32bit, for os::has_allocatable_memory_limit
+// Helper, on 32bit, for os::allocatable_memory_limit
```
Pre-existing: paying for mmap only to later pay for the same mmap again seems overly costly, seeing that there is no guarantee anyway that the later mmap will also succeed.

jsikstro commented:
I agree with you both, @albertnetymk and @tstuefe, that `commit_memory_limit` and `reserve_memory_limit` are good names and convey the intention of the functions better.

> ```
> // reserve_memory_limit() returns an upper limit beyond which os::reserve_memory
> // calls are guaranteed to fail. It is not guaranteed reserving smaller sizes will succeed,
> // however.
> ```
> 
> , and vice versa for commit_memory_limit.

I also like having this kind of comment above.

jsikstro commented:
Pushed three new commits where I rename `address_space_limit` to `reserve_memory_limit`, rename `allocatable_memory_limit` to `commit_memory_limit`, and reordered the header along with adding some describing comments for both functions.

jsikstro commented:
With some suggestions from @albertnetymk, I've changed `commit_memory_limit` to return size_t, just as `reserve_memory_limit`. I think this is reasonable since `commit_memory_limit` should always be bounded by what `reserve_memory_limit` returns anyway, so having both return size_t is good.

I also refactored some of code so that it fits better with the fact that `commit_memory_limit` always returns a value.

tstuefe reviewed (approved):
Good.
